File: src/src/main/resources/mmlspark/_JSONOutputParser.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class _JSONOutputParser(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaTransformer):
    """


    Args:

        dataType (object): format to parse the column to
        inputCol (str): The name of the input column
        outputCol (str): The name of the output column
        postProcessor (object): optional transformation to postprocess json output
    """

    @keyword_only
    def __init__(self, dataType=None, inputCol=None, outputCol=None, postProcessor=None):
        super(_JSONOutputParser, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.JSONOutputParser")
        self._cache = {}
        self.dataType = Param(self, "dataType", "dataType: format to parse the column to")
        self.inputCol = Param(self, "inputCol", "inputCol: The name of the input column")
        self.outputCol = Param(self, "outputCol", "outputCol: The name of the output column")
        self.postProcessor = Param(self, "postProcessor", "postProcessor: optional transformation to postprocess json output", generateTypeConverter("postProcessor", self._cache, complexTypeConverter))
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, dataType=None, inputCol=None, outputCol=None, postProcessor=None):
        """
        Set the (keyword only) parameters

        Args:

            dataType (object): format to parse the column to
            inputCol (str): The name of the input column
            outputCol (str): The name of the output column
            postProcessor (object): optional transformation to postprocess json output
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setDataType(self, value):
        """

        Args:

            dataType (object): format to parse the column to

        """
        self._set(dataType=value)
        return self


    def getDataType(self):
        """

        Returns:

            object: format to parse the column to
        """
        return self.getOrDefault(self.dataType)


    def setInputCol(self, value):
        """

        Args:

            inputCol (str): The name of the input column

        """
        self._set(inputCol=value)
        return self


    def getInputCol(self):
        """

        Returns:

            str: The name of the input column
        """
        return self.getOrDefault(self.inputCol)


    def setOutputCol(self, value):
        """

        Args:

            outputCol (str): The name of the output column

        """
        self._set(outputCol=value)
        return self


    def getOutputCol(self):
        """

        Returns:

            str: The name of the output column
        """
        return self.getOrDefault(self.outputCol)


    def setPostProcessor(self, value):
        """

        Args:

            postProcessor (object): optional transformation to postprocess json output

        """
        self._set(postProcessor=value)
        return self


    def getPostProcessor(self):
        """

        Returns:

            object: optional transformation to postprocess json output
        """
        return self._cache.get("postProcessor", None)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.JSONOutputParser"

    @staticmethod
    def _from_java(java_stage):
        module_name=_JSONOutputParser.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".JSONOutputParser"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/JSONOutputParser.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using Microsoft.Spark.Utils;

using SynapseML.Dotnet.Utils;


namespace Synapse.ML.Io.Http
{
    /// <summary>
    /// <see cref="JSONOutputParser"/> implements JSONOutputParser
    /// </summary>
    public class JSONOutputParser : JavaTransformer, IJavaMLWritable, IJavaMLReadable<JSONOutputParser>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.io.http.JSONOutputParser";

        /// <summary>
        /// Creates a <see cref="JSONOutputParser"/> without any parameters.
        /// </summary>
        public JSONOutputParser() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="JSONOutputParser"/> with a UID that is used to give the
        /// <see cref="JSONOutputParser"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public JSONOutputParser(string uid) : base(s_className, uid)
        {
        }

        internal JSONOutputParser(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets dataType value for <see cref="dataType"/>
        /// </summary>
        /// <param name="dataType">
        /// format to parse the column to
        /// </param>
        /// <returns> New JSONOutputParser object </returns>
        public JSONOutputParser SetDataType(DataType value) =>
            WrapAsJSONOutputParser(Reference.Invoke("setDataType",
            DataType.FromJson(Reference.Jvm, value.Json)));
        
        /// <summary>
        /// Sets inputCol value for <see cref="inputCol"/>
        /// </summary>
        /// <param name="inputCol">
        /// The name of the input column
        /// </param>
        /// <returns> New JSONOutputParser object </returns>
        public JSONOutputParser SetInputCol(string value) =>
            WrapAsJSONOutputParser(Reference.Invoke("setInputCol", (object)value));
        
        /// <summary>
        /// Sets outputCol value for <see cref="outputCol"/>
        /// </summary>
        /// <param name="outputCol">
        /// The name of the output column
        /// </param>
        /// <returns> New JSONOutputParser object </returns>
        public JSONOutputParser SetOutputCol(string value) =>
            WrapAsJSONOutputParser(Reference.Invoke("setOutputCol", (object)value));
        
        /// <summary>
        /// Sets postProcessor value for <see cref="postProcessor"/>
        /// </summary>
        /// <param name="postProcessor">
        /// optional transformation to postprocess json output
        /// </param>
        /// <returns> New JSONOutputParser object </returns>
        public JSONOutputParser SetPostProcessor(JavaTransformer value) =>
            WrapAsJSONOutputParser(Reference.Invoke("setPostProcessor", (object)value));

        
        /// <summary>
        /// Gets dataType value for <see cref="dataType"/>
        /// </summary>
        /// <returns>
        /// dataType: format to parse the column to
        /// </returns>
        public DataType GetDataType()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getDataType");
            string json = (string)jvmObject.Invoke("json");
            return DataType.ParseDataType(json);
        }
        
        
        /// <summary>
        /// Gets inputCol value for <see cref="inputCol"/>
        /// </summary>
        /// <returns>
        /// inputCol: The name of the input column
        /// </returns>
        public string GetInputCol() =>
            (string)Reference.Invoke("getInputCol");
        
        
        /// <summary>
        /// Gets outputCol value for <see cref="outputCol"/>
        /// </summary>
        /// <returns>
        /// outputCol: The name of the output column
        /// </returns>
        public string GetOutputCol() =>
            (string)Reference.Invoke("getOutputCol");
        
        
        /// <summary>
        /// Gets postProcessor value for <see cref="postProcessor"/>
        /// </summary>
        /// <returns>
        /// postProcessor: optional transformation to postprocess json output
        /// </returns>
        public JavaTransformer GetPostProcessor()
        {
            JvmObjectReference jvmObject = (JvmObjectReference)Reference.Invoke("getPostProcessor");
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaTransformer),
                "s_className");
            JvmObjectUtils.TryConstructInstanceFromJvmObject(
                jvmObject,
                classMapping,
                out JavaTransformer instance);
            return instance;
        }

        
        /// <summary>
        /// Loads the <see cref="JSONOutputParser"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="JSONOutputParser"/> was saved to</param>
        /// <returns>New <see cref="JSONOutputParser"/> object, loaded from path.</returns>
        public static JSONOutputParser Load(string path) => WrapAsJSONOutputParser(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<JSONOutputParser> Read() =>
            new JavaMLReader<JSONOutputParser>((JvmObjectReference)Reference.Invoke("read"));

        private static JSONOutputParser WrapAsJSONOutputParser(object obj) =>
            new JSONOutputParser((JvmObjectReference)obj);

        
    }
}

        